singularity-opac/constants/constants.hpp

```diff
@@ -26,239 +26,186 @@
 };
 
-struct CGS {
-  static constexpr double length = 1.e2;                    // centimeter
-  static constexpr double mass = 1.e3;                      // gram
-  static constexpr double time = 1.;                        // second
-  static constexpr double temperature = 1.;                 // Kelvin
-  static constexpr double current = 1.e-1;                  // Biot
-  static constexpr double charge = 2.997924580e9;           // Statcoulomb
-  static constexpr double capacitance = 8.9831483395497e11; // Statfarad
-  static constexpr double angle = 1.;                       // Radian
+struct BaseSICODATA2010 {
+  // Avogadro constant
+  static constexpr Real avogadro = 6.02214129e23;
+  // Fine structure constant
+  static constexpr Real fine_structure = 7.2973525698e-3;
+  // Planck constant
+  static constexpr Real planck = 6.62606957e-34;
+  // Boltzmann constant
+  static constexpr Real boltzmann = 1.380648800e-23;
+  // Elementary charge
+  static constexpr Real elementary_charge = 1.602176565e-19;
+  // Speed of light (exact)
+  static constexpr Real speed_of_light = 2.99792458e8;
+  // Gravitational constant
+  static constexpr Real gravitational_constant = 6.67384e-11;
+  // Standard acceleration of gravity
+  static constexpr Real acceleration_from_gravity = 9.80665;
+  // Electron rest mass
+  static constexpr Real electron_mass = 9.10938291e-31;
+  // Proton rest mass
+  static constexpr Real proton_mass = 1.672621777e-27;
+  // Neutron rest mass
+  static constexpr Real neutron_mass = 1.674927351e-27;
+  // Atomic mass unit
+  static constexpr Real atomic_mass_unit = 1.660538921e-27;
+  // Permeability of free space
+  static constexpr Real vacuum_permeability = 4.0 * M_PI * 1.0e-7;
+  // Fermi coupling constant (Units of erg^{-2})
+  static constexpr Real fermi_coupling_constant = 4.543791885043567014;
+  // Axial-vector coupling
+  static constexpr Real axial_vector_coupling_constant = -1.23;
 };
 
-// Defines and encapsulates physical and mathematical constants in a purely
-// constexpr way, with support for multiple unit systems.
-template <typename UNITSYSTEM>
-class PhysicalConstants {
+struct UnitConversionDefault {
+  static constexpr Real length = 1.;
+  static constexpr Real mass = 1.;
+  static constexpr Real time = 1.;
+  static constexpr Real temperature = 1.;
+  static constexpr Real current = 1.;
+  static constexpr Real charge = 1.;
+};
+
+struct UnitConversionSIToCGS {
+  static constexpr Real length = 1.e2;          // centimeter
+  static constexpr Real mass = 1.e3;            // gram
+  static constexpr Real time = 1.;              // second
+  static constexpr Real temperature = 1.;       // Kelvin
+  static constexpr Real current = 1.e-1;        // Biot
+  static constexpr Real charge = 2.997924580e9; // Statcoulomb
+};
+
+template <typename BASE = BaseSICODATA2010,
+          typename CONVERSION = UnitConversionSIToCGS>
+struct PhysicalConstants {
```

Comment:
This is a pretty drastic reimaging of the physical constants framework. Basically now it takes in a class of base constants AND a unit conversion factor. This allows for more flexibility, including fake units that still maintain some internal consistency (e.g. the radiation constant is properly derived from planck's constant). Not all unit systems can be achieved by a conversion factor (mainly for test problems that change physical constants to achieve certain solution behaviors). 

My main issue with this, I think, is that the opacity classes (opacity, thermal distribution, and mean opacity) are all now templated on the unit system. I couldn't think of an easy way to ensure consistency between all of these -- maybe we need a new container class that holds all three of these and doles out template parameters appropriately?

Reply:
Hmmm. Yeah I feel a little uncertain about that. A couple thoughts that may or may not address the issue:
1. Can the template parameter default to constants in CGS, so you don't have to explicitly enter a template param unless you know what you're doing?
2. Should the `Opac` object be templated on units... e.g., 
```C++
template<typename Units>
using Opac = opac_variant<GrayOpacity<Units>, TophatOpacity<Units>>
```
etc.
3. How does this interact with tabular data which *cannot* be templated on units?
4. What if the set of physical constants was an optional parameter passed into the constructor?

Reply:
`1.` was my idea originally, but I would sometimes get errors saying c++14 couldn't handle that level of template deduction. I didn't really understand the problem except that it seems like nested template parameters with default arguments doesn't work in c++14. 

`2.` is interesting...

`3.` where units aren't used (e.g. tabular data and gray (scale-free) opacities) the units are just ignored. Well, the units are always used in the thermal distributions. Maybe for tabular data with specific units I should add a static assert so one must use the appropriate (CGS) units.

`4.` I'd like to keep the constants `constexpr` if at all possible

Reply:
Returning to this:
`2.` and `3.` actually does only explicitly combining certain physical constants units with certain opacities in the current `using Opacity = ` not solve this problem? Someone downstream would have to manually type e.g. `SpinerOpacity<FermiDiracDistributionNoMu<3, PhysicalConstantsUnity>, PhysicalConstantsUnity>` in their code to get to an invalid state.

Reply:
Ah, I see what you're saying, @brryan, yes that's true. I'd stilll like to return to `1` if possible though... It would be ideal if the user didn't *have* to specify units. I don't know.